flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java

```diff
@@ -71,4 +71,41 @@ public void testKeepFirstRowWithGenerateRetraction() throws Exception {
 		testHarness.close();
 	}
 
+	@Test
+	public void tesKeepFirstRowWithStateTtl() throws Exception{
+		MiniBatchDeduplicateKeepFirstRowFunction func = new MiniBatchDeduplicateKeepFirstRowFunction(typeSerializer, minTime.toMilliseconds());
+		OneInputStreamOperatorTestHarness<BaseRow, BaseRow> testHarness = createTestHarness(func);
+		testHarness.setup();
+		testHarness.open();
+		testHarness.processElement(record("book", 1L, 12));
+		testHarness.processElement(record("book", 2L, 11));
+		// output is empty because bundle not trigger yet.
+		Assert.assertTrue(testHarness.getOutput().isEmpty());
+		testHarness.processElement(record("book", 1L, 13));
+
+		Thread.sleep(30);
```

Comment:
Do we have any other way to test this, I don't think `Thread.sleep` is a good idea.

Reply:
Does `testHarness.setProcessingTime()` work in this case? cc @Myasuka , can we use `testHarness.setProcessingTime` to test the expiration of state ttl?

Reply:
state ttl use TtlTimeProvider.DEFAULT to get current process time, testHarness.setProcessingTime doesn't work, whether it is appropriate to use TimeUnit.MILLISECONDS.sleep(30) ? Do you have any better advice?

Reply:
Unfortunately, current we always use `TtlTimeProvider.DEFAULT` to create keyed state backend. One way to walk around this:

1. Define a new `MockTtlTimeProvider` which could set current time outside.
1. Define a `TTLMemoryStateBackend` extended from `MemoryStateBackend` which would ignore the given `TtlTimeProvider.DEFAULT` when creating keyed state backend, but use previous newly defined  `MockTtlTimeProvider`.
1. Set the `TTLMemoryStateBackend` to the newly `KeyedOneInputStreamOperatorTestHarness` to use our customized state backend.

Then you could set a new time to `MockTtlTimeProvider` to ensure time has passed time to live.

Reply:
@wuchong , @Myasuka At the moment, KeyedStateBackend created in `StreamTaskStateInitializerImpl#keyedStateBackend` method using `TtlTimeProvider.DEFAULT` explicitly, didn't support pass TtlTimeProvider to it, even though we implement `TTLMemoryStateBackend `, we also can't  set `MockTtlTimeProvider ` to KeyedStateBackend. Unless, we modify relavant code in `StreamTaskStateInitializerImpl`, but I dare not change the code because I can't evaluate the impact, what do you think?

Reply:
Hi @lsyldliu , I think a better way is to make  `StreamTaskStateInitializerImpl` accept a `TtlTimeProvider` constructor parameter instead of hard code `TtlTimeProvider.DEFAULT` in the implementation. Then, we can pass-in the `MockTtlTimeProvider` in `AbstractStreamOperatorTestHarness#createStreamTaskStateManager`. 

What do you think @Myasuka ? Could you help on this?

Reply:
@wuchong , sure I could help on this to let `TtlTimeProvider` configurable by opening another issue.

Reply:
Thanks @Myasuka . 
@lsyldliu , we may need to hold this PR until `StreamTaskStateInitializerImpl` supports custom `TtlTimeProvider`.

Reply:
OK